internal/middleware/account.go

```diff
@@ -3,29 +3,54 @@ package middleware
 import (
 	"net/http"
 
+	"github.com/RHEnVision/provisioning-backend/internal/cache"
 	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
 	"github.com/RHEnVision/provisioning-backend/internal/dao"
+	"github.com/RHEnVision/provisioning-backend/internal/models"
 )
 
 func AccountMiddleware(next http.Handler) http.Handler {
 	fn := func(w http.ResponseWriter, r *http.Request) {
-		rhId := ctxval.Identity(r.Context())
 		logger := ctxval.Logger(r.Context())
-		accDao, err := dao.GetAccountDao(r.Context())
-		if err != nil {
-			logger.Error().Err(err).Msg("Failed to initialize connection to fetch Account info")
-			http.Error(w, http.StatusText(500), 500)
-			return
+		rhId := ctxval.Identity(r.Context())
+		orgID := rhId.Identity.OrgID
+		accountNumber := rhId.Identity.AccountNumber
+		cacheKey := cache.AccountKey{
+			OrgID:         orgID,
+			AccountNumber: accountNumber,
 		}
-		acc, err := accDao.GetOrCreateByIdentity(r.Context(), rhId.Identity.OrgID, rhId.Identity.AccountNumber)
-		if err != nil {
-			logger.Error().Err(err).Msgf("Failed to fetch account by org_id=%s/account=%s", rhId.Identity.OrgID, rhId.Identity.AccountNumber)
-			http.Error(w, http.StatusText(500), 500)
-			return
+
+		var foundAccount *models.Account
+		if cachedAccount, ok := cache.FindAccountId(r.Context(), cacheKey); ok {
+			// account found in cache
+			foundAccount = cachedAccount
+			logger.Trace().Int64("account", foundAccount.ID).Msg("Account cache hit")
+		} else {
+			// account not found in cache
+			accDao, err := dao.GetAccountDao(r.Context())
+			if err != nil {
+				logger.Error().Err(err).Msg("Failed to initialize connection to fetch Account info")
+				http.Error(w, http.StatusText(500), 500)
+				return
+			}
+
+			account, err := accDao.GetOrCreateByIdentity(r.Context(), orgID, accountNumber)
+			if err != nil {
+				logger.Error().Err(err).Msgf("Failed to fetch account by org_id=%s/account=%s", orgID, accountNumber)
+				http.Error(w, http.StatusText(500), 500)
+				return
+			}
+
+			cache.SetAccountId(r.Context(), cacheKey, account)
+			foundAccount = account
 		}
 
-		newLogger := logger.With().Str("org_id", acc.OrgID).Str("account_number", acc.AccountNumber.String).Logger()
-		ctx := ctxval.WithAccountId(r.Context(), acc.ID)
+		newLogger := logger.With().
+			Int64("account", foundAccount.ID).
+			Str("org_id", foundAccount.OrgID).
+			Str("acc_n", foundAccount.AccountNumber.String).
```

Comment:
It is not clear from the naming what is the difference between the account and acc_n
Can you specify the difference between them in the name?

Reply:
@lzap What do you think?

Reply:
Yeah, account is our internal account id, acc_n. Let me do it this way:

* account_id - our ID
* org_id
* account_number